### PR TITLE
[PLA-2023] Improve metadata caching

### PR DESCRIPTION
**Pull request by enjinabner**
### **PR Type**
enhancement, configuration changes

___

### **Description**
- Added a new command `SyncAttributeMetadata` to sync attribute metadata to cache, improving performance.
- Introduced caching in `MetadataService` with methods `fetchAndCache` and `getCache` for efficient metadata retrieval.
- Updated `Token` model to utilize cached metadata instead of fetching it directly.
- Added configuration for attribute metadata syncing, including a `data_chunk_size` setting.
- Registered the new command in the `CoreServiceProvider`.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Configuration changes</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>enjin-platform.php</strong><dd><code>Add configuration for attribute metadata syncing</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

config/enjin-platform.php

<li>Added configuration for attribute metadata syncing.<br> <li> Introduced <code>sync_metadata</code> array with <code>data_chunk_size</code> setting.<br>

</details>

  </td>
  <td><a href="https://github.com/enjin/platform-core/pull/260/files#diff-e687a9c8af991f77587526ea6983e8c280929f8ccf6b989eb1eb83c0d8f4d220">+12/-0</a>&nbsp; &nbsp; </td>

</tr>                    
</table></td></tr><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>SyncAttributeMetadata.php</strong><dd><code>Implement SyncAttributeMetadata command for caching</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

src/Commands/SyncAttributeMetadata.php

<li>Created new command <code>SyncAttributeMetadata</code>.<br> <li> Command syncs attribute metadata to cache.<br> <li> Utilizes progress bar for tracking sync progress.<br>

</details>

  </td>
  <td><a href="https://github.com/enjin/platform-core/pull/260/files#diff-d0a91a2d6496f6ceebd58a4872043263d29b9851c2ceeaba60f3f56141125ad4">+53/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>CoreServiceProvider.php</strong><dd><code>Register SyncAttributeMetadata command in service provider</code></dd></summary>
<hr>

src/CoreServiceProvider.php

- Registered `SyncAttributeMetadata` command.

</details>

  </td>
  <td><a href="https://github.com/enjin/platform-core/pull/260/files#diff-7239ef674a3243e4c9895004a5f8ad86f9264922f450c554605b3a1c7fe02d62">+2/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>Token.php</strong><dd><code>Use cached metadata in Token model</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

src/Models/Laravel/Token.php

<li>Changed metadata fetching to use cache.<br> <li> Replaced <code>MetadataService::fetch</code> with <code>MetadataService::getCache</code>.<br>

</details>

  </td>
  <td><a href="https://github.com/enjin/platform-core/pull/260/files#diff-c43aa5a65eba10e52488cbe7124d7bfed85857e528af165379a505edd28c5597">+3/-3</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>MetadataService.php</strong><dd><code>Enhance MetadataService with caching capabilities</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

src/Services/Database/MetadataService.php

<li>Added caching functionality for metadata.<br> <li> Introduced <code>fetchAndCache</code> and <code>getCache</code> methods.<br> <li> Implemented cache key generation.<br>

</details>

  </td>
  <td><a href="https://github.com/enjin/platform-core/pull/260/files#diff-81af9cad135cb3d44832130138aba169cfa1ee5407d78da7aae4f0d573e9ac3d">+33/-0</a>&nbsp; &nbsp; </td>

</tr>                    
</table></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**: Comment `/help "your question"` on any pull request to receive relevant information

### CODE REVIEW

**Review by leonardocustodio — COMMENTED** [review]
I've not tested it yet but I wonder if wouldn't be better to dispatch the requests as jobs instead of using a worker.

And maybe use a "lower priority" queue

**Comment by enjinabner** [comment]
> I've not tested it yet but I wonder if wouldn't be better to dispatch the requests as jobs instead of using a worker.
> 
> And maybe use a "lower priority" queue

I'm considering using concurrency in laravel instead of queue so I can still show the actual progress in the progress bar.

**Comment by enjinabner** [comment]
> > I've not tested it yet but I wonder if wouldn't be better to dispatch the requests as jobs instead of using a worker.
> > And maybe use a "lower priority" queue
> 
> I'm considering using concurrency in laravel instead of queue so I can still show the actual progress in the progress bar. Throwing it to the queue is good but I wont know what actually happened in the queue.

**Comment by leonardocustodio** [comment]
My concern is making the requets too fast, concurrency could make it even worse.

I'm thinking about this scenario:

- Let's say a lot of tokens/users store their metadata in the same hosting provider. Or even could be their own provider but they have lots of token...

Making too many requests at once could get us flagged as being a "bot/spammer" and rate-limit us or even get us blocked.

I don't think there is a "urgency" to complete this "job" as fast as possible. Making something more slow but steady could be a better approach, also not sure that we do need to follow the progress.

Maybe could we store something like: "metadata_last_updated_at" ?

Just throwing some ideas here, let me know what you think

**Comment by enjinabner** [comment]
> My concern is making the requets too fast, concurrency could make it even worse.
> 
> I'm thinking about this scenario:
> 
> * Let's say a lot of tokens/users store their metadata in the same hosting provider. Or even could be their own provider but they have lots of token...
> 
> Making too many requests at once could get us flagged as being a "bot/spammer" and rate-limit us or even get us blocked.
> 
> I don't think there is a "urgency" to complete this "job" as fast as possible. Making something more slow but steady could be a better approach, also not sure that we do need to follow the progress.
> 
> Maybe could we store something like: "metadata_last_updated_at" ?
> 
> Just throwing some ideas here, let me know what you think

That's a good point, however aside from performance I wanna make sure that the user will always get the updated metadata, that's why its important to finish it fast. I've also considered `metadata_last_updated_at` but I still need to read the url and compare the data, so its kinda pointless.

Maybe we can consider a webhook were adaptors need to call to make sure that platform will know that the metadata is updated, otherwise we'll sync them in a scheduled manner.

**Comment by leonardocustodio** [comment]
Wouldn't a mutation `refreshMetadata` solve the issue? Just like in the indexer?
If the adopter knows the metadata has changed, he can call it with the collectionId / tokenId and it will refresh instantly

**Comment by enjinabner** [comment]
> Wouldn't a mutation `refreshMetadata` solve the issue? Just like in the indexer? If the adopter knows the metadata has changed, he can call it with the collectionId / tokenId and it will refresh instantly

Yea, creating a mutation is fine as well. As long as adaptor has the ability to refresh it themselves. But we need to mention this in the docs

**Review by leonardocustodio — DISMISSED** [review]

File: src/GraphQL/Schemas/Primary/Substrate/Mutations/RefreshMetadataMutation.php [hunk 1]
```diff
@@ -73,6 +76,16 @@ public function resolve(
         SerializationServiceInterface $serializationService,
         MetadataService $metadataService,
     ): mixed {
+        $key = 'RefreshMetadataMutation'
+                . ':' . Arr::get($args, 'collectionId')
+                . ':' . $this->encodeTokenId($args);
```
Comment (leonardocustodio) [inline src/GraphQL/Schemas/Primary/Substrate/Mutations/RefreshMetadataMutation.php]: There is a PlatformCache key class that should be used for cache key values

**Review by leonardocustodio — CHANGES_REQUESTED** [review]

**Review by leonardocustodio — DISMISSED** [review]

**Review by leonardocustodio — APPROVED** [review]

**Review by leonardocustodio — APPROVED** [review]

**Pull request merged**